test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -423,22 +424,133 @@ class test_ucp_tag_limits : public test_ucp_tag {
-    bool m_test_offload;
+    bool    m_test_offload;
+    size_t  m_min_rndv;
+
+    static void check_short_thresh(const ucp_memtype_thresh_t &thresh,
+                                   size_t cfg_thresh)
+    {
+        EXPECT_LE(thresh.memtype_on + 1, cfg_thresh);
+        EXPECT_LE(thresh.memtype_off + 1, cfg_thresh);
+    }
+
+    void check_rndv_startup_config(size_t exp_rndv_thresh,
+                                   size_t exp_rndv_intra_thresh,
+                                   size_t exp_rndv_inter_thresh)
+    {
+        ucp_context_config_t *cfg = &sender().worker()->context->config.ext;
+
+        EXPECT_EQ(exp_rndv_thresh, cfg->rndv_thresh);
+        EXPECT_EQ(exp_rndv_intra_thresh, cfg->rndv_intra_thresh);
+        EXPECT_EQ(exp_rndv_inter_thresh, cfg->rndv_inter_thresh);
+    }
+
+    void check_tag_rndv_v2(size_t cfg_thresh)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+
+        if (m_test_offload) {
+            /* If configured threshold is less than min_rndv, then expect exact
+             * min_rndv limit for short messages */
+            if (cfg_thresh < m_min_rndv) {
+                EXPECT_EQ(cfg->tag.offload.max_eager_short.memtype_on + 1,
+                          m_min_rndv);
+                EXPECT_EQ(cfg->tag.offload.max_eager_short.memtype_off + 1,
+                          m_min_rndv);
+            } else {
+                check_short_thresh(cfg->tag.offload.max_eager_short, cfg_thresh);
+            }
+        } else {
+            check_short_thresh(cfg->tag.max_eager_short, cfg_thresh);
+        }
+    }
+
+    void check_am_rndv_v2(size_t cfg_thresh)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+
+        check_short_thresh(cfg->am_u.max_eager_short, cfg_thresh);
+        check_short_thresh(cfg->am_u.max_reply_eager_short, cfg_thresh);
+    }
+
+    void check_ep_proto_rndv(size_t cfg_thresh, bool expect_rndv)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+        ucp_proto_select_t *proto_select = &cfg->proto_select;
+        size_t msg_length = cfg_thresh;
+        ucp_proto_select_elem_t value;
+
+        if (m_test_offload) {
+            /* There is a lower bound for rndv threshold with tag offload.
+             * We should not send messages smaller than this limit */
+            msg_length = ucs_max(m_min_rndv, msg_length);
+        }
+
+        kh_foreach_value(proto_select->hash, value, {
+            /* Find index of the corresponding ucp_proto_threshold_elem_t
+             * to handle the given message size */
+            unsigned idx = 0;
+            const ucp_proto_threshold_elem_t *th;
+            for (; msg_length > value.thresholds[idx].max_msg_length; ++idx) {
+                th = &value.thresholds[idx];
+                /* Assert no rndv before expected limit */
+                EXPECT_EQ(nullptr, strstr(th->proto_config.proto->name, "rndv"));
+            }
+            th = &value.thresholds[idx];
```

Comment:
Minor: It seems like test doesn't required any information from `ucp_proto_threshold_elem_t` except `proto_config::name` and `proto_config::cfg_thresh` so there is no need to use logic based on threshold entity.

Pls consider this:
```suggestion
            const ucp_proto_config_t *proto_config;
            for (; msg_length > value.thresholds[idx].max_msg_length; ++idx) {
                proto_config = &value.thresholds[idx].proto_config;
                /* Assert no rndv before expected limit */
                EXPECT_EQ(nullptr, strstr(proto_config->proto->name, "rndv"));
            }
            proto_config = &value.thresholds[idx].proto_config;
```

Or even this:
```suggestion
            std::string proto_name;
            size_t cfg_thresh;
            
            for (; msg_length > value.thresholds[idx].max_msg_length; ++idx) {
                proto_name = value.thresholds[idx].proto_config.proto->name;
                /* Assert no rndv before expected limit */
                EXPECT_EQ(std::string::npos, proto_name.find("rndv"));
            }
            proto_name       = value.thresholds[idx].proto_config.proto->name;
            proto_cfg_thresh = value.thresholds[idx].proto_config.cfg_thresh;
```

Reply:
sure, makes sense